azblob/chunkwriting.go

```diff
@@ -0,0 +1,214 @@
+package azblob
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+)
+
+// blockWriter provides methods to upload blocks that represent a file to a server and commit them.
+// This allows us to provide a local implementation that fakes the server for hermetic testing.
+type blockWriter interface {
+	StageBlock(context.Context, string, io.ReadSeeker, LeaseAccessConditions, []byte) (*BlockBlobStageBlockResponse, error)
+	CommitBlockList(context.Context, []string, BlobHTTPHeaders, Metadata, BlobAccessConditions) (*BlockBlobCommitBlockListResponse, error)
+}
+
+// copyFromReader copies a source io.Reader to blob storage using concurrent uploads.
+// TODO(someone): The existing model provides a buffer size and buffer limit as limiting factors.  The buffer size is probably
+// useless other than needing to be above some number, as the network stack is going to hack up the buffer over some size. The
+// max buffers is providing a cap on how much memory we use (by multiplying it times the buffer size) and how many go routines can upload
+// at a time.  I think having a single max memory dial would be more efficient.  We can choose an internal buffer size that works
+// well, 4 MiB or 8 MiB, and autoscale to as many goroutines within the memory limit. This gives a single dial to tweak and we can
+// choose a max value for the memory setting based on internal transfers within Azure (which will give us the maximum throughput model).
+// We can even provide a utility to dial this number in for customer networks to optimize their copies.
+func copyFromReader(ctx context.Context, from io.Reader, to blockWriter, o UploadStreamToBlockBlobOptions) (*BlockBlobCommitBlockListResponse, error) {
+	o.defaults()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	copy := &copier{
```

Comment:
>copy [](start = 1, length = 4)

NItpick: `copy` is a predefined keyword you're shadowing. perhaps activeCopier may be preferable?

Reply:
That is an excellent point.  I remember that I meant to go back and do that (clearly, did not).  

How about "cp"?

Reply:
Sure, that's alright by me.

---
In reply to: [410440223](https://github.com/Azure/azure-storage-blob-go/pull/167#discussion_r410440223) [](ancestors = 410440223)